docs/docs/reference/extend/extension-methods.md

```diff
@@ -0,0 +1,146 @@
+
+```scala
+case class Circle(x: Double, y: Double, radius: Double)
+
+extension CircleOps for Circle {
+  def circumference: Double = this.radius * math.Pi * 2
+}
+```
+
+The extension adds a method `circumference` to values of class `Circle`. Like regular methods, extension methods can be invoked with infix `.`:
+
+```scala
+  val circle = Circle(0, 0, 1)
+  circle.circumference
+```
+
+### Meaning of `this`
+
+Inside an extension method, the name `this` stands for the receiver on which the
+method is applied when it is invoked. E.g. in the application of `circle.circumference`,
+the `this` in the body of `circumference` refers to `circle`. As usual, `this` can be elided, so we could also have defined `CircleOps` like this:
+
+```scala
+extension CircleOps for Circle {
+  def circumference: Double = radius * math.Pi * 2
+}
+```
+
+### Scope of Extensions
+
+Extensions can appear anywhere in a program; there is no need to co-define them with the types they extend. Extension methods are available wherever their defining extension is in scope.  Extensions can be inherited or imported like normal definitions.
+
+### Extended Types
+
+An extension can add methods to arbitrary types. For instance, the following
+clause adds a `longestStrings` extension method to a `Seq[String]`:
+
+```scala
+extension StringOps for Seq[String] {
+  def longestStrings: Seq[String] = {
+    val maxLength = map(_.length).max
+    filter(_.length == maxLength)
+  }
+}
+```
+
+### Generic Extensions
+
+The previous example extended a specific instance of a generic type. It is also possible
+to extend a generic type by adding type parameters to an extension:
+
+```scala
+extension ListOps[T] for List[T] {
+  def second: T = tail.head
+}
+```
+
+or:
+
+
+```scala
+extension ListListOps[T] for List[List[T]] {
+  def flattened: List[T] = foldLeft[List[T]](Nil)(_ ++ _)
+}
+```
+
+### Bounded Generic Extensions
+
+It is possible to use bounds for the type parameters of an extension. But subtype and context bounds are supported:
+
+```scala
+extension ShapeListOps[T <: Shape] for List[T] {
+  def totalArea = map(_.area).sum
+}
+
+extension OrdSeqOps[T : math.Ordering] for Seq[T] {
+  def indexOfLargest  = zipWithIndex.maxBy(_._1)._2
+  def indexOfSmallest = zipWithIndex.minBy(_._1)._2
+}
+```
+
+### Implicit Parameters for Type Patterns
+
+The standard translation of context bounds expands the bound in the last example to an implicit _evidence_ parameter of type `math.Ordering[T]`. It is also possible to give evidence parameters explicitly. The following example is equivalent to the previous one:
+
+```scala
+extension OrdSeqOps[T](implicit ev: math.Ordering[T]) for Seq[T] {
+  def indexOfLargest  = this.zipWithIndex.maxBy(_._1)._2
+  def indexOfSmallest = this.zipWithIndex.minBy(_._1)._2
+}
+```
+
+There can be only one parameter clause following a type pattern and it must be implicit. As usual, one can combine context bounds and implicit evidence parameters.
```

Comment:
In principle, it seems the only needed restriction is that no explicit parameter lists are allowed; otherwise, any combination of implicit parameter lists/type parameter lists could make sense (as soon as the relevant bugs are fixed elsewhere).